File: packages/types/src/contribution.ts
/**
 * Contribution types for the GSD Community Hub.
 *
 * These mirror the on-chain ContributionLeaf schema and related
 * off-chain metadata for the contribution tracking system.
 */

/**
 * Core contribution data matching the on-chain ContributionLeaf fields.
 *
 * Field sizes match the Rust struct serialization:
 * - developer: 32 bytes (Pubkey)
 * - taskRef: 32 bytes (SHA-256 hash of task identifier)
 * - verificationScore: 2 bytes (u16, 0-10000 = 0.00%-100.00%)
 * - timestamp: 8 bytes (i64, unix timestamp)
 * - contentHash: 32 bytes (SHA-256 hash of contribution content)
 *
 * Total serialized: 106 bytes
 */
export interface ContributionData {
  /** Wallet public key of the developer (base58 string, 32 bytes on-chain) */
  developer: string;
  /** SHA-256 hash of the task identifier (hex string, 32 bytes on-chain) */
  taskRef: string;
  /** Verification score: 0-10000 representing 0.00%-100.00% */
  verificationScore: number;
  /** Unix timestamp in seconds when the contribution was recorded */
  timestamp: number;
  /** SHA-256 hash of off-chain contribution content (hex string, 32 bytes on-chain) */
  contentHash: string;
}

/**
 * Full contribution record including off-chain metadata.
 * Extends ContributionData with database fields.
 */
export interface ContributionRecord extends ContributionData {
  /** Unique database identifier */
  id: string;
  /** Human-readable task title or reference */
  taskTitle: string;
  /** Brief description of the contribution */
  description: string;
  /** Leaf hash from the Merkle tree (hex string) */
  leafHash: string;
  /** Index in the Merkle tree, if appended */
  leafIndex?: number;
  /** Solana transaction signature that recorded this on-chain */
  txSignature?: string;
  /** Database creation timestamp */
  createdAt: Date;
  /** Database update timestamp */
  updatedAt: Date;
}

/**
 * Contribution score summary for a developer.
 * Maps to the score fields in the on-chain DeveloperProfile account.
 */
export interface ContributionScore {
  /** Number of verified tasks completed */
  tasksCompleted: number;
  /** Sum of all verification scores (for average calculation) */
  totalVerificationScore: bigint;
  /** Days active since first contribution */
  timeActiveDays: number;
  /** Contribution score scaled by 1e6 for fixed-point precision */
  contributionScore: bigint;
  /** Unix timestamp of first contribution, 0 = none */
  firstContributionAt: number;
  /** Unix timestamp of most recent contribution */
  lastContributionAt: number;
  /** Score formula version */
  scoreVersion: number;
}

File: programs/gsd-hub/src/state/contribution.rs
use anchor_lang::prelude::*;
use sha2::{Digest, Sha256};

/// Leaf data for the contribution Merkle tree.
///
/// This is NOT a PDA account -- it is serialized, hashed, and appended
/// as a leaf to the spl-account-compression concurrent Merkle tree.
/// The full serialized data (106 bytes) is also emitted via spl-noop
/// so off-chain indexers can reconstruct contributions from tx logs.
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug, PartialEq)]
pub struct ContributionLeaf {
    /// Wallet that made the contribution (32 bytes)
    pub developer: Pubkey,
    /// SHA-256 hash of the task identifier (32 bytes)
    pub task_ref: [u8; 32],
    /// Verification score: 0-10000 representing 0.00%-100.00% (2 bytes)
    pub verification_score: u16,
    /// Unix timestamp of when the contribution was recorded (8 bytes)
    pub timestamp: i64,
    /// SHA-256 hash of off-chain contribution content (32 bytes)
    pub content_hash: [u8; 32],
}
// Total serialized: 32 + 32 + 2 + 8 + 32 = 106 bytes

impl ContributionLeaf {
    /// Serialized size of the leaf data.
    pub const SIZE: usize = 106;

    /// Serialize the leaf and hash it for Merkle tree insertion.
    ///
    /// Returns a 32-byte SHA-256 hash suitable for use as a concurrent Merkle tree leaf.
    pub fn to_leaf_hash(&self) -> [u8; 32] {
        let data = AnchorSerialize::try_to_vec(self).expect("ContributionLeaf serialization");
        let mut hasher = Sha256::new();
        hasher.update(&data);
        hasher.finalize().into()
    }
}
